api/utils/keypaths/keypaths.go

```diff
@@ -0,0 +1,282 @@
+	fileExtSSHCert = "-cert.pub"
+	// fileExtPub is the extension of a file where a public key is stored.
+	fileExtPub = ".pub"
+	// appDirSuffix is the suffix of a sub-directory where app TLS certs are stored.
+	appDirSuffix = "-app"
+	// db DirSuffix is the suffix of a sub-directory where db  TLS certs are stored.
+	dbDirSuffix = "-db"
+	// kubeDirSuffix is the suffix of a sub-directory where kube TLS certs are stored.
+	kubeDirSuffix = "-kube"
+)
+
+// Here's the file layout of all these keypaths.
+// ~/.tsh/							 --> default base directory
+// ├── known_hosts                   --> trusted certificate authorities (their keys) in a format similar to known_hosts
+// └── keys
+//    ├── one.example.com            --> Proxy hostname
+//    │   ├── certs.pem              --> TLS CA certs for the Teleport CA
+//    │   ├── foo                    --> RSA Private Key for user "foo"
+//    │   ├── foo.pub                --> Public Key
+//    │   ├── foo-x509.pem           --> TLS client certificate for Auth Server
+//    │   ├── foo-ssh                --> SSH certs for user "foo"
+//    │   │   ├── root-cert.pub      --> SSH cert for Teleport cluster "root"
+//    │   │   └── leaf-cert.pub      --> SSH cert for Teleport cluster "leaf"
+//    │   ├── foo-kube               --> Kubernetes certs for user "foo"
+//    │   │   ├── root               --> Kubernetes certs for Teleport cluster "root"
+//    │   │   │   ├── kubeA-x509.pem --> TLS cert for Kubernetes cluster "kubeA"
+//    │   │   │   └── kubeB-x509.pem --> TLS cert for Kubernetes cluster "kubeB"
+//    │   │   └── leaf               --> Kubernetes certs for Teleport cluster "leaf"
+//    │   │       └── kubeC-x509.pem --> TLS cert for Kubernetes cluster "kubeC"
+//    │   └── foo-db                 --> Database access certs for user "foo"
+//    │       ├── root               --> Database access certs for cluster "root"
+//    │       │   ├── dbA-x509.pem   --> TLS cert for database service "dbA"
+//    │       │   └── dbB-x509.pem   --> TLS cert for database service "dbB"
+//    │       └── leaf               --> Database access certs for cluster "leaf"
+//    │           └── dbC-x509.pem   --> TLS cert for database service "dbC"
+//    └── two.example.com
+//        ├── certs.pem
+//        ├── bar
+//        ├── bar.pub
+//        ├── bar-x509.pem
+//        └── bar-ssh
+//            └── clusterA-cert.pub
+
+// KeyDir returns the path to the keys directory.
+//
+// <baseDir>/keys
+func KeyDir(baseDir string) string {
+	return filepath.Join(
+		baseDir,
+		sessionKeyDir,
+	)
+}
+
+// KnownHostsPath returns the path to the known hosts file.
+//
+// <baseDir>/known_hosts
+func KnownHostsPath(baseDir string) string {
+	return filepath.Join(baseDir, fileNameKnownHosts)
+}
+
+// ProxyKeyDir returns the path to the proxy's keys directory.
+//
+// <baseDir>/keys/<proxy>
+func ProxyKeyDir(baseDir, proxy string) string {
+	return filepath.Join(
+		KeyDir(baseDir),
```

Comment:
YMMV: It's probably premature optimisation, but something is nagging at me in having to allocate a whole string for an intermediate result, which is then immediately thrown away when it could just be embedded straight in the call to `filepath.Join()`.

Reply:
You're right, I'll expand it out instead of using intermediary functions.